satellite/overlay/nodeselectioncache.go

```diff
@@ -0,0 +1,195 @@
+// Copyright (C) 2019 Storj Labs, Incache.
+// See LICENSE for copying information.
+
+package overlay
+
+import (
+	"context"
+	"math/rand"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+)
+
+// CacheDB implements the database for overlay node selection cache
+//
+// architecture: Database
+type CacheDB interface {
+	// SelectAllStorageNodesUpload returns all nodes that qualify to store data, organized as reputable nodes and new nodes
+	SelectAllStorageNodesUpload(ctx context.Context, selectionCfg NodeSelectionConfig) (reputable, new []*SelectedNode, err error)
+}
+
+// CacheConfig is a configuration for overlay node selection cache.
+type CacheConfig struct {
+	Staleness time.Duration `help:"how stale the node selection cache can be" releaseDefault:"3m" devDefault:"5m"`
+}
+
+// NodeSelectionCache keeps a list of all the storage nodes that are qualified to store data
+// We organize the nodes by if they are reputable or a new node on the network.
+// The cache will sync with the nodes table in the database and get refreshed once the staleness time has past.
+type NodeSelectionCache struct {
+	log             *zap.Logger
+	db              CacheDB
+	selectionConfig NodeSelectionConfig
+	staleness       time.Duration
+
+	mu   sync.RWMutex
+	data *state
+}
+
+type state struct {
+	lastRefresh time.Time
+
+	mu             sync.RWMutex
+	reputableNodes []*SelectedNode
+	newNodes       []*SelectedNode
+}
+
+// NewNodeSelectionCache creates a new cache that keeps a list of all the storage nodes that are qualified to store data
+func NewNodeSelectionCache(log *zap.Logger, db CacheDB, staleness time.Duration, config NodeSelectionConfig) *NodeSelectionCache {
+	return &NodeSelectionCache{
+		log:             log,
+		db:              db,
+		staleness:       staleness,
+		selectionConfig: config,
+		data:            &state{},
+	}
+}
+
+// Init populates the cache with all of the reputableNodes and newNode nodes
+// that qualify to upload data from the nodes table in the overlay database
+func (cache *NodeSelectionCache) Init(ctx context.Context) (err error) {
```

Comment:
Oh, I just realized that you don't even need the init, since during the first call to GetNodes it will update the cache.

Reply:
i think it might be nice to init the cache when the api starts up. but i could be talked in to only doing it during GetNodes. although i would like to try to reduce the number of times we refresh from within the GetNodes method since its blocking. thoughts?

Reply:
I mean, it might take some time from init and to first call to GetNodes... so it might do that anyways. I don't think it will make a perf difference, however, I would remove this mainly because it will be less code.

Reply:
ok i removed Init method, but replaced with public Refresh method to use in tests